test/test_framework.py

```diff
@@ -130,6 +132,25 @@ def on_foo(self, event):
 
         self.assertEqual(obs.seen, ["on_any:foo", "on_foo:foo", "on_any:bar"])
 
+    def test_on_pre_commit_emitted(self):
+        framework = self.create_framework()
+
+        class PreCommitObserver(Object):
+            def __init__(self, parent, key):
+                super().__init__(parent, key)
+                self.seen = []
+
+            def on_pre_commit(self, event):
+                self.seen.append(type(event))
```

Comment:
1) Setting new state in `PreCommitObserver.on_pre_commit` will result in an observer being created with `StoredStateData.on_pre_commit` as a handler, however, `StoredStateData.on_pre_commit` will *not* see this event because the observation is set up after `pre_commit` is emitted. So we will get `AttributeError: attribute 'mydata' is not stored`.

2) Modifying existing state in on_pre_commit on a custom object will not work either:

- `StoredStateData.on_pre_commit` -> data gets snapshotted;
- `PreCommitObserver.on_pre_commit` -> the same data gets modified -> never gets saved because `StoredStateData.on_pre_commit` have already ran at this point.

So, currently this is not the right place to set or modify state. Do we want it to be?

Reply:
I introduced another event called `commit` now.

So, `pre_commit` is for developers to persist data in their objects before commit, `commit` is for `StoredStateData` to call `save_snapshot`.